### DISCUSSION TITLE
Document select keeps the pdf with the same size.

### DISCUSSION THREAD

**Discussion by caiocesarrm** [pymupdf/PyMuPDF#884] [Q&A] [answered]
Hi!
I have a pdf (a book with 373 pages). I select one page using doc.select([75]) and save.

```
doc = fitz.open(f'{base_path}[Chollet]-Deep_Learning_with_Python.pdf')
doc.select([75])
doc.save(f'{base_path}page.pdf', clean=True, garbage=4, deflate=True)
```

the page saved: [page.pdf](https://github.com/pymupdf/PyMuPDF/files/5923629/page.pdf)

![image](https://user-images.githubusercontent.com/41218670/106853506-91ac9180-6698-11eb-9467-19d6d0ea7e1d.png)

After i save the file, the pdf have almost the same size, having only one page. 
i believe that this is because the content is still there. Any thoughts on how can i solve this?

![image](https://user-images.githubusercontent.com/41218670/106853742-f536bf00-6698-11eb-9690-56dfe73d1e71.png)

PyMuPDF 1.18.6.
Built for Python 3.6 on linux (64-bit).

**Comment by JorjMcKie** [comment]
There may be a number of reasons for this - even apart from a potential PyMuPDF bug 😉.

Garbage collection can only remove objects which are no longer referenced by anything whatsoever ... and that may be all sorts of this.
Other possible explanations are potentially lots of / large embedded files.
Also, that single page may still use large fonts for its text or show big images ...
The important thing to notice about ``select``: apart from removing pages, the PDF is not changed otherwise.

Here are a few steps to get closer to an explanation:
1. Before selecting, do a ``page.clean_contents(sanitize=True)`` (or ``cleanContents(sanitize=True)`` for an older version).
   - cleaning removes any unneeded references from the page definition ... many PDFs are created in that unfortunate way ...
2. Do that page cleaning before saving.
3. Make a new PDF and insert that one page into it:

```python
new = fitz.open()
new.insert_pdf(doc, from_page=75, to_page=75)  # or insertPDF(...)
new.save(...)
```

**Comment by JorjMcKie** [comment, marked as answer]
Found the reason for this unexspected size behaviour:
The page contains a `/B` entry
```
B array (Optional; PDF 1.1; recommended if the page contains article beads) An ar-
ray of indirect references to article beads appearing on the page (see Sec-
tion 8.3.2, “Articles”; see also implementation note 37 in Appendix H).
The beads are listed in the array in natural reading order.
```
In this case it references annotations, which in turn reference other pages, and yet other pages, etc.
Result: the full PDF is still there - only the page pointers have been reduced to that one page.

So you have the option 3 from previous post - which gives you below 70 KB PDF size.
**_OR_** with functions new in v1.18.7, you can - half-manually - remove (= set to PDF ``null``) that entry:

```python
>>> import fitz
>>> doc=fitz.open("page.pdf")
>>> page=doc[0]
>>> print(doc.xref_get_keys(page.xref))
('Annots', 'B', 'Contents', 'CropBox', 'MediaBox', 'Parent', 'Resources', 'Rotate', 'Type')
>>> doc.xref_set_key(page.xref, "B", "null")
>>> page.clean_contents()
>>> doc.save("x.pdf",garbage=4,deflate=True)
```

Which results in a PDF of 66 KB.

Semi-manual might also mean
```python
for key in doc.xref_get_keys(page.xref):
    if key not in ('Annots', 'Contents', 'CropBox', 'MediaBox', 'Parent', 'Resources', 'Rotate', 'Type'):
        doc.xref_set_key(page.xref, key, "null")
```
So this would only keep PDF keys also included by the `doc.insert_pdf` method.
